### DISCUSSION TITLE
EntityCallback Support in MongoDB Panache Library similar to spring data ?

### DISCUSSION THREAD

**Discussion by nitinty** [quarkusio/quarkus#43989] [Q&A]
Hi Team,

Does the MongoDB Panache library provide EntityCallback support similar to what Spring offers with callbacks like AfterConvertCallback, BeforeConvertCallback, or AfterSaveCallback?

Our requirement is to perform encryption and decryption logic before/after conversion when saving or fetching records from the database.

Could you please advise how we can achieve this. ?

[a comment by an automated account is not shown]

**Reply by nitinty** [reply]
@FroMage @geoand 
Can you please help on this ?

**Reply by geoand** [reply]
I have not been involved in any Mongo Panache work (AFAIR at least), so I'm probably not the best person to comment on this

**Reply by nitinty** [reply]
Sorry to bother you, If possible can you please tag the right person, If I missed anyone here ?

**Reply by geoand** [reply]
@loicmathieu might have ideas

**Comment by loicmathieu** [comment]
No, we don't provide these kind of callbacks.
But we automatically configure any `CommandListener.` you add in the classpath so you may be able to do this using a listener.

**Reply by nitinty** [reply]
@loicmathieu 
The CommandListener will be invoked for all entity types whenever any DB operation is performed, regardless of whether we intend to update a specific entity or not. I’ll need to implement a dirty check for my specific document/entity, which could negatively impact performance. 
Could you please suggest if there’s a way to configure the listener for a specific entity or document only ?

**Reply by FroMage** [reply]
It might be nice to support common callbacks for ORM and Mongo, I have a vague memory that it's not the first time someone asks for a callback...

**Reply by nitinty** [reply]
Yeah, that would be great and would also give us the flexibility to use it as needed.

**Reply by FroMage** [reply]
See https://github.com/quarkusio/quarkus/issues/11164 for someone asking for this already in the past, and https://github.com/quarkusio/quarkus/discussions/31794 for another type of annotation (timestamps) that would have made sense to share support with ORM as well.

**Reply by nitinty** [reply]
I also raised #41612 for auditing some time back.

**Reply by nitinty** [reply]
@loicmathieu 
I tried the listener approach as shown below, but it didn’t work as expected. The document was saved with the original value instead of the updated one.

```java 

import com.mongodb.event.CommandFailedEvent;
import com.mongodb.event.CommandListener;
import com.mongodb.event.CommandStartedEvent;
import com.mongodb.event.CommandSucceededEvent;
import io.quarkus.logging.Log;
import lombok.extern.slf4j.Slf4j;
import org.bson.BsonArray;
import org.bson.BsonDocument;
import org.bson.BsonString;

@Slf4j
public class TestEntityCommandListener implements CommandListener {
    @Override
    public void commandStarted(CommandStartedEvent event) {
        Log.info("CommandStartedEvent " + event.getCommand());
        BsonArray documents = (BsonArray) event.getCommand().get("documents");
        BsonArray updatedDocumentArray = new BsonArray(
                documents.stream().map(mongoDocument -> {
                    BsonDocument updatedDoc = new BsonDocument();
                    // Iterate through key-value pairs only once
                    mongoDocument.asDocument().forEach((key, value) -> {
                        if (value instanceof BsonString) {
                            // Add "_updated" only to BsonString values
                            updatedDoc.put(key, new BsonString(((BsonString) value).getValue() + "_updated"));
                        } else {
                            updatedDoc.put(key, value);
                        }
                    });
                    return updatedDoc;
                }).toList()
        );
        // Here set the updated document in the event
        event.getCommand().put("documents", updatedDocumentArray);
    }

    @Override
    public void commandSucceeded(CommandSucceededEvent event) {
        Log.info("CommandSucceededEvent " + event.getCommandName());
    }

    @Override
    public void commandFailed(CommandFailedEvent event) {
        Log.info("CommandFailedEvent " + event.getCommandName());
    }
}

```

<img width="1155" alt="Screenshot 2024-10-23 at 12 58 56 PM" src="https://github.com/user-attachments/assets/8eaaf8c2-cb84-4145-8150-bbb08d55eec4">

**Reply by xstefank** [reply]
I also tried this and I can confirm that I can't modify the value in the inserted document. We need to do it before insert so we need to provide a separate API to do it.

**Reply by xstefank** [reply]
@loicmathieu do you agree that we can provide callbacks in our mongo Panache extension? I will log the issue then.

**Reply by loicmathieu** [reply]
The first step is to open an issue describing what your "dream" API will be.
Then, if you have some implementation idea you can give them in the issue description but if we want to provide callback for both MongoDB and Hibernate with panache we will need to sync to defines the behaviour in both implémentation (and for the reactive variant).

**Reply by nitinty** [reply]
@loicmathieu @FroMage 

I have opened issue #44066 and updated it with the required details. I would appreciate it if you could take a look. Please feel free to update or suggest any better approaches!

@xstefank
